codegen/src/main/twirl/templates/PlayJavaServer/RouterUsingActions.scala.txt

```diff
@@ -0,0 +1,80 @@
+ *@
+
+@(service: akka.grpc.gen.javadsl.Service, powerApis: Boolean)
+
+@akka.grpc.gen.Constants.DoNotEditComment
+package @service.packageName;
+
+import akka.actor.ActorSystem;
+import akka.grpc.GrpcServiceException;
+import akka.grpc.internal.PlayRouterHelper$;
+import akka.grpc.internal.PlayRouterUsingActions;
+import akka.grpc.javadsl.GrpcExceptionHandler;
+import akka.http.javadsl.model.HttpRequest;
+import akka.http.javadsl.model.HttpResponse;
+import akka.japi.Function;
+import akka.stream.Materializer;
+import play.api.mvc.EssentialAction;
+import play.api.mvc.RequestHeader;
+import play.api.mvc.PlayBodyParsers;
+import play.api.mvc.DefaultActionBuilder;
+
+import scala.Function1;
+import scala.NotImplementedError;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+@defining(if (powerApis) service.name + "PowerApi" else service.name) { serviceName =>
+  /**
+   * Abstract base class for implementing @{serviceName} and using as a play Router
+   */
+  public abstract class Abstract@{serviceName}Router extends PlayRouterUsingActions implements @{serviceName} {
+    private final Function<ActorSystem, Function<Throwable, io.grpc.Status>> eHandler;
+    private final ActorSystem system;
+    private final PlayBodyParsers parsers;
+    private final DefaultActionBuilder actionBuilder;
+    @{
+      val (streamingInputMethods: Seq[String], unaryInputMethods: Seq[String]) = service.methods.partition(_.inputStreaming) match {
+        case (streaming, unary) => (streaming.map(_.grpcName), unary.map(_.grpcName))
+      }
+      s"""
+      private final List<String> streamingInputMethods = Arrays.asList(${streamingInputMethods.mkString("\"", "\", \"", "\"")});
+      private final List<String> unaryInputMethods = Arrays.asList(${unaryInputMethods.mkString("\"", "\", \"", "\"")});
+      """
+    }
+
+    public Abstract@{serviceName}Router(Materializer mat, ActorSystem system, PlayBodyParsers parsers, DefaultActionBuilder actionBuilder) {
+      this(mat, system, parsers, actionBuilder, GrpcExceptionHandler.defaultMapper());
+    }
+
+    public Abstract@{serviceName}Router(Materializer mat, ActorSystem system, PlayBodyParsers parsers, DefaultActionBuilder actionBuilder, Function<ActorSystem, Function<Throwable, io.grpc.Status>> eHandler) {
+      super(mat, @{service.name}.name, parsers, actionBuilder);
+      this.eHandler = eHandler;
+      this.system = system;
+      this.parsers = parsers;
+      this.actionBuilder = actionBuilder;
+    }
+
+    @@Override
+    public Function1<RequestHeader, EssentialAction> createHandler(String serviceName, Materializer mat) {
+      return new Function1<RequestHeader, EssentialAction>() {
+        public EssentialAction apply(RequestHeader reqOuter) {
+          Function<HttpRequest, CompletionStage<HttpResponse>> handler = @{serviceName}HandlerFactory.create(Abstract@{serviceName}Router.this, serviceName, mat, eHandler, system);
```

Comment:
If I'm reading this correctly, this would now be called for every request. It looks like that would not be necessary, and it could be created once outside of the function. Would that work?

Reply:
No, this is only called once during initialization. See [here](https://github.com/akka/akka-grpc/pull/534/files#diff-7e9bbe2f58513cfce8675cfe00c583d7R44).

It shouldn't be public though, will change that.

Reply:
Sure, `createHandler` is only called once, but line 66 is inside the function that is returned by `createHandler`, and that is called for every request - unless I'm misreading this somehow, it could be moved to between line 64 to only create the handler once? (haven't tried though)

Reply:
Ah right, I was looking at line 64. Will fix.

Reply:
Fixed